File: Assignment5_Sangines/CNNByAMParallel2/CNNByAM/CNNLayer.cs
using CNNAM;
using MatrixLibByAM;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CNNByAM
{
    [Serializable]
    class CNNLayer
    {
        public CNNLayer(int numFeatureMaps, int numPrevLayerFeatureMaps, int inputSize,
            int kernelSize, PoolingType poolingType, ActivationType activationType, int batchSize)
        {
            this.batchSize = batchSize;
            this.ConvolSums = new Matrix[batchSize,numFeatureMaps];

            this.kernelSize = kernelSize;
            this.numFeatureMaps = numFeatureMaps;
            this.numPrevLayerFeatureMaps = numPrevLayerFeatureMaps;

            this.ConvolResults = new Matrix[batchSize,numPrevLayerFeatureMaps, numFeatureMaps];

            int convOutputSize = inputSize - kernelSize + 1;

            for (int i = 0; i < batchSize; i++)
                for (int j = 0; j < numFeatureMaps; j++)
                    ConvolSums[i,j] = new Matrix(convOutputSize, convOutputSize);
            Kernels = new Matrix[numPrevLayerFeatureMaps, numFeatureMaps];
            KernelGrads = new Matrix[numPrevLayerFeatureMaps, numFeatureMaps];
            InitMatrix2DArray(ref Kernels,numPrevLayerFeatureMaps, numFeatureMaps,kernelSize);
            InitMatrix2DArray(ref KernelGrads, numPrevLayerFeatureMaps, numFeatureMaps, kernelSize);

            InitializeKernels();
            for (int i = 0; i < numFeatureMaps; i++)
            {
                FeatureMap fmp = new FeatureMap(convOutputSize, poolingType, activationType,batchSize);
                FeatureMapList.Add(fmp);
            }
        }

        int kernelSize;
        // will do prev maps * numFeatureMap convolutions
        // for the first CNN layer, it will do 1xnumFeatureMaps convolutions
        // for subsequent layers, it will do prevLayerFeatureMapsxnumFeatureMaps
        // convolutions. Even though first CNN layer does not require a 2-D
        // array of matrices to store kernels, gradients, we will still use
        // 1xnumFeaturemaps arrays so that we can generalize to any number of 
        // CNN layers 
        int batchSize = 0;
        public Matrix[,] Kernels = null;
        public Matrix[,] KernelGrads = null;
        public Matrix[,,] ConvolResults = null;
        public Matrix[,] ConvolSums = null;
        int numFeatureMaps;
        int numPrevLayerFeatureMaps;
        public List<FeatureMap> FeatureMapList = new List<FeatureMap>();
        Random rand = new Random((int)(DateTime.Now.Ticks));  // 5000 = seed for repeatable results

        void InitializeKernels()  // initialize all kernels
        {
            for (int i = 0; i < Kernels.GetLength(0); i++)
                for (int j = 0; j < Kernels.GetLength(1); j++)
                    InitializeKernel(Kernels[i, j]);
        }

        public void Evaluate(List<Matrix>PrevLayerOutputList, int batchIndex)
        {   // inputs come from outputs of previous layer
            // do convolutions with outputs of feature maps from previous layer
            for (int p = 0; p < numPrevLayerFeatureMaps; p++)
            {
                for (int q = 0; q < numFeatureMaps; q++)
                //Parallel.For(0, numFeatureMaps, (q) =>
                {
                    ConvolResults[batchIndex,p, q] = PrevLayerOutputList[p].Convolution(Kernels[p, q]);
                }
            }
            // add convolution results
            for (int q = 0; q < FeatureMapList.Count; q++)
            //Parallel.For(0, FeatureMapList.Count, (q) => 
            {
                ConvolSums[batchIndex,q].Clear();
                for (int p = 0; p < PrevLayerOutputList.Count; p++)
                {
                    ConvolSums[batchIndex,q] = ConvolSums[batchIndex,q] + ConvolResults[batchIndex,p, q];
                }
            }
            // evaluate each feature map i.e., perform activation after adding bias
            for(int i = 0; i < FeatureMapList.Count;i++)
            {
                FeatureMapList[i].Evaluate(ConvolSums[batchIndex,i],batchIndex);
            }
        }

        void InitMatrix2DArray(ref Matrix[,] Mat, int dim1, int dim2, int matrixSize)
        {
            Mat = new Matrix[dim1,dim2];
            for (int i = 0; i < dim1; i++)
            {
                for (int j = 0; j < dim2; j++)
                {
                    Mat[i, j] = new Matrix(matrixSize, matrixSize);
                }
            }
        }

        void InitializeKernel(Matrix kernel)
        {
            for (int i = 0; i < kernel.Rows; i++)
            {
                for (int j = 0; j < kernel.Cols; j++)
                {
                    double num = rand.NextDouble();
                    if (num < 0.5)
                        kernel.D[i][j] = rand.NextDouble() * 0.1;  // D for data
                    else
                        kernel.D[i][j] = rand.NextDouble() * -0.1;
                }
            }
        }
    }
}

File: Assignment5_Sangines/CNN_Sangines/CNN_Sangines/CNNLayer.py
import numpy as np



class CNNLayer(object):

    def __init__(self, numFeatureMaps, numPrevLayerFeatureMaps, inputSize, kernelSize, poolingType, activationType, batchSize):
        self.batchSize = batchSize #Images passing
        self.convolSum = np.zeros((batchSize,numFeatureMaps)) #Each image (in batch) will have one Sum for each FeatureMap we want. 

        self.kernelSizes = kernelSize
        self.numFeatureMaps = numFeatureMaps #Features Maps we want on the current Layer
        self.numPrevLayerFeatureMaps = numPrevLayerFeatureMaps #Feature Maps coming from previous Layer.

        self.convolResults = np.zeros((batchSize,numPrevLayerFeatureMaps, numFeatureMaps)) #Each img in the Batch will have an array 
                                                                                           #Each array represents an output Feature from pev layer
                                                                                           #Each prev feature will have sizes equal to the Features needed in current Layer
        self.featureMapList = [] #List of feature maps in the actual Layer
        for i in range(self.numFeatureMaps): #Looping though all the feature maps needed for the Layer
            featureMp = FeatureMap(convOutputSize, poolingType, activationType, batchSize) #creating Feature Map
            self.featureMapList.append(featureMp)

        self.convOutputSize = inputSize - kernelSize +1 #Sizes after applying Convolution
        self.kernels = np.zeros((numPrevLayerFeatureMaps, numFeatureMaps, kernelSize, kernelSize)) #Kernels matrix will have sizes of PrevFeatures x CurrentFeatures
        self.kernelsGrads = np.zeros((numPrevLayerFeatureMaps, numFeatureMaps, kernelSize, kernelSize))


        #In each img in batch from ConvolSum, we will store X matrixes of size (convolRes x convolRes) where X = Total Features we want in the Layer
        for i in range(batchSize):
            for j in range(numFeatureMaps):
                self.convolSum[i, j] = np.zeros((self.convOutputSize,self.convOutputSize))
        
        #Initializing Kernels 
        self.kernels = InitializeKernel(numPrevLayerFeatureMaps,numFeatureMaps, kernelSize)
        #self.kernels = InitMatrixesInArray(numPrevLayerFeatureMaps,numFeatureMaps, kernelSize)
        #self.kernelsGrads = InitMatrixesInArray(numPrevLayerFeatureMaps,numFeatureMaps, kernelSize)

    #creating empty matrixes for kernels
    #def InitMatrixesInArray(self, dimPrev, dimActual, subDim):
        #tempKern = [np.zeros((subDim,subDim)) for i in range(dimPrev*dimActual)]
        #return np.asarray(tempKern)

    #Initialize Kernels
    #def InitializeKernels(self):
        #self.kernels = [InitializeKernel(self.kernels[i] for i in range(len(self.kernels)))]
        #for i in range(len(self.kernels)):
        #    InitializeKernel(self.kernels[i])

    def InitializeKernel(self, dim1, dim2, kSize):
        temp = np.zeros((dim1, dim2, kSize, kSize))
        for i in range(len(temp[0])):
            for j in range(len(temp[0,0])):
                for a in range(len(temp[0,0,0])):
                    temp[i,j,a] = [np.random.uniform(0,1) * 0.1 if np.random.uniform(0,1)<0.5 else np.random.uniform(0,1) * -0.1 for item in temp[i,j,a]]

                #num = np.random.uniform(0,1)
                #if num<0.5:
                    #temp[i,j] = np.random.uniform(0,1) * 0.1
                #else:
                    #temp[i,j] = np.random.uniform(0,1) * -0.1
        return temp

    def Evaluate(self, PrevLayerOutput, batchIndex):
        #Doing convolutions with the features (outputs) from the Prev Layer
        for i in range(self.numPrevLayerFeatureMaps): #Each iteration is a PrevFeature Map 
            for j in range(self.numFeatureMaps): #Number of Feature Maps we want in current Layer
                self.convolResults[batchIndex, i, q] = PrevLayerOutput[i].Convolution(kernels[i,j])




            



